src/daft-plan/src/optimization/rules/push_down_projection.rs

```diff
@@ -52,6 +52,93 @@ impl PushDownProjection {
             return Ok(new_plan);
         }
 
+        // Next, check if the upstream is another projection we can merge with.
+        // This is possible iff the upstream projection's computation-required columns
+        // are each only used once in this downstream projection.
+        if let LogicalPlan::Project(upstream_projection) = upstream_plan.as_ref() {
+            // Get all the computation-required columns from the upstream projection.
+            let upstream_computations = upstream_projection
+                .projection
+                .iter()
+                .flat_map(|e| {
+                    e.input_mapping().map_or_else(
+                        // None means computation required -> Some(colname)
+                        || Some(e.name().unwrap().to_string()),
+                        // Some(computation not required) -> None
+                        |_| None,
+                    )
+                })
+                .collect::<IndexSet<_>>();
+
+            // For each of them, make sure they are used only once in this downstream projection.
+            let mut exprs_to_walk: Vec<Arc<Expr>> = projection
+                .projection
+                .iter()
+                .map(|e| e.clone().into())
+                .collect();
+
+            let mut upstream_computations_used = IndexSet::new();
+            let mut okay_to_merge = true;
+
+            while !exprs_to_walk.is_empty() {
+                exprs_to_walk = exprs_to_walk
+                    .iter()
+                    .flat_map(|expr| {
+                        // If it's a reference for a column that requires computation,
+                        // record it.
+                        if let Expr::Column(name) = expr.as_ref() {
+                            if upstream_computations.contains(name.as_ref()) {
+                                okay_to_merge = okay_to_merge
+                                    && upstream_computations_used.insert(name.to_string())
+                            }
+                        };
+                        if okay_to_merge {
+                            expr.children()
+                        } else {
+                            vec![]
+                        }
+                    })
+                    .collect();
+            }
```

Comment:
Nit: This expression tree walk technically won't short-circuit until a particular level of the expression tree has been fully traversed, but I think that should be fine.